File: cpp/src/topology.cpp
#include "hics/topology.hpp"

#include <algorithm>
#include <cmath>
#include <limits>
#include <queue>
#include <unordered_map>

namespace hics {

double edge_latency_us(const EdgeAttributes& attrs, uint64_t size_bytes,
                       double utilization) {
    utilization = std::clamp(utilization, 0.0, 0.99);
    double residual = 1.0 - utilization;
    double transfer_s = static_cast<double>(size_bytes) / 1e9 /
                        (attrs.bandwidth_gbps * std::pow(residual, attrs.contention_alpha));
    return attrs.latency_us + transfer_s * 1e6;
}

uint32_t TopologyGraph::add_vertex(EndpointId endpoint) {
    vertices_.push_back(endpoint);
    return static_cast<uint32_t>(vertices_.size() - 1);
}

uint32_t TopologyGraph::add_edge(uint32_t src, uint32_t dst, EdgeAttributes attrs) {
    edges_.push_back({src, dst, attrs});
    return static_cast<uint32_t>(edges_.size() - 1);
}

uint32_t TopologyGraph::vertex_of(const EndpointId& ep) const {
    for (size_t i = 0; i < vertices_.size(); ++i) {
        if (vertices_[i] == ep) return static_cast<uint32_t>(i);
    }
    return UINT32_MAX;
}

void TopologyGraph::precompute_paths(uint32_t k, uint32_t max_hops) {
    path_cache_.clear();
    const uint32_t n = static_cast<uint32_t>(vertices_.size());
    for (uint32_t src = 0; src < n; ++src) {
        for (uint32_t dst = 0; dst < n; ++dst) {
            if (src == dst) continue;
            path_cache_[path_key(src, dst)] = yen_k_shortest(src, dst, k, max_hops);
        }
    }
}

const std::vector<std::vector<uint32_t>>& TopologyGraph::paths(uint32_t src,
                                                               uint32_t dst) const {
    static const std::vector<std::vector<uint32_t>> empty;
    auto it = path_cache_.find(path_key(src, dst));
    return it != path_cache_.end() ? it->second : empty;
}

std::vector<std::vector<uint32_t>> TopologyGraph::yen_k_shortest(
    uint32_t src, uint32_t dst, uint32_t k, uint32_t max_hops) const {
    // Simplified Yen's k-shortest paths using BFS with hop limit
    struct State {
        uint32_t vertex;
        std::vector<uint32_t> edge_path;
        double cost;
        bool operator>(const State& o) const { return cost > o.cost; }
    };

    std::vector<std::vector<uint32_t>> result;
    std::priority_queue<State, std::vector<State>, std::greater<State>> pq;
    pq.push({src, {}, 0.0});

    while (!pq.empty() && result.size() < k) {
        auto [v, path, cost] = pq.top();
        pq.pop();
        if (v == dst) {
            result.push_back(path);
            continue;
        }
        if (path.size() >= max_hops) continue;

        for (size_t ei = 0; ei < edges_.size(); ++ei) {
            if (edges_[ei].src_vertex != v) continue;
            // Avoid cycles
            bool cycle = false;
            for (auto e : path) {
                if (edges_[e].dst_vertex == edges_[ei].dst_vertex) {
                    cycle = true;
                    break;
                }
            }
            if (cycle) continue;

            auto new_path = path;
            new_path.push_back(static_cast<uint32_t>(ei));
            double edge_cost = edges_[ei].attrs.latency_us +
                               1e6 / edges_[ei].attrs.bandwidth_gbps;
            pq.push({edges_[ei].dst_vertex, new_path, cost + edge_cost});
        }
    }
    return result;
}

TopologyGraph TopologyGraph::build_h100_node(uint32_t node_id, uint32_t num_gpus) {
    TopologyGraph g;
    std::vector<uint32_t> gpu_verts;

    for (uint32_t i = 0; i < num_gpus; ++i) {
        gpu_verts.push_back(g.add_vertex({node_id, i, EndpointType::GPU}));
    }
    uint32_t cpu = g.add_vertex({node_id, 0, EndpointType::CPU});
    uint32_t cxl = g.add_vertex({node_id, 0, EndpointType::CXLController});
    uint32_t ib = g.add_vertex({node_id, 0, EndpointType::IBPort});

    auto nvlink = default_fabric_attrs(FabricType::NVLink);
    auto pcie = default_fabric_attrs(FabricType::PCIe);
    auto cxl_attrs = default_fabric_attrs(FabricType::CXL);
    auto ib_attrs = default_fabric_attrs(FabricType::InfiniBand);

    // NVLink all-to-all within node
    for (uint32_t i = 0; i < num_gpus; ++i) {
        for (uint32_t j = 0; j < num_gpus; ++j) {
            if (i != j) g.add_edge(gpu_verts[i], gpu_verts[j], nvlink);
        }
    }
    // PCIe GPU-CPU
    for (auto gv : gpu_verts) {
        g.add_edge(gv, cpu, pcie);
        g.add_edge(cpu, gv, pcie);
        g.add_edge(gv, cxl, cxl_attrs);
        g.add_edge(cxl, gv, cxl_attrs);
        g.add_edge(gv, ib, ib_attrs);
        g.add_edge(ib, gv, ib_attrs);
    }
    g.add_edge(cpu, ib, pcie);
    g.add_edge(ib, cpu, pcie);

    return g;
}

TopologyGraph TopologyGraph::build_cluster(uint32_t num_nodes, uint32_t gpus_per_node) {
    TopologyGraph cluster;
    std::vector<std::vector<uint32_t>> node_gpus(num_nodes);
    std::vector<uint32_t> node_ib(num_nodes);

    for (uint32_t n = 0; n < num_nodes; ++n) {
        auto node = build_h100_node(n, gpus_per_node);
        uint32_t base = static_cast<uint32_t>(cluster.vertices().size());
        for (const auto& v : node.vertices()) cluster.add_vertex(v);
        for (const auto& e : node.edges()) {
            cluster.add_edge(e.src_vertex + base, e.dst_vertex + base, e.attrs);
        }
        for (uint32_t g = 0; g < gpus_per_node; ++g) {
            node_gpus[n].push_back(base + g);
        }
        node_ib[n] = base + gpus_per_node + 2;  // GPU×N + CPU + CXL + IB
    }

    auto ib = default_fabric_attrs(FabricType::InfiniBand);
    for (uint32_t i = 0; i < num_nodes; ++i) {
        for (uint32_t j = i + 1; j < num_nodes; ++j) {
            cluster.add_edge(node_ib[i], node_ib[j], ib);
            cluster.add_edge(node_ib[j], node_ib[i], ib);
        }
    }
    return cluster;
}

}  // namespace hics

File: rust/src/topology.rs
use crate::types::*;
use std::collections::{HashMap, HashSet, VecDeque};

#[derive(Debug, Clone)]
pub struct GraphEdge {
    pub src_vertex: u32,
    pub dst_vertex: u32,
    pub attrs: EdgeAttributes,
}

pub struct TopologyGraph {
    vertices: Vec<EndpointId>,
    edges: Vec<GraphEdge>,
    path_cache: HashMap<(u32, u32), Vec<Vec<u32>>>,
}

impl TopologyGraph {
    pub fn new() -> Self {
        Self {
            vertices: Vec::new(),
            edges: Vec::new(),
            path_cache: HashMap::new(),
        }
    }

    pub fn add_vertex(&mut self, endpoint: EndpointId) -> u32 {
        self.vertices.push(endpoint);
        (self.vertices.len() - 1) as u32
    }

    pub fn add_edge(&mut self, src: u32, dst: u32, attrs: EdgeAttributes) -> u32 {
        self.edges.push(GraphEdge {
            src_vertex: src,
            dst_vertex: dst,
            attrs,
        });
        (self.edges.len() - 1) as u32
    }

    pub fn vertices(&self) -> &[EndpointId] {
        &self.vertices
    }

    pub fn edges(&self) -> &[GraphEdge] {
        &self.edges
    }

    pub fn vertex_of(&self, ep: &EndpointId) -> Option<u32> {
        self.vertices
            .iter()
            .position(|v| v == ep)
            .map(|i| i as u32)
    }

    pub fn paths(&self, src: u32, dst: u32) -> &[Vec<u32>] {
        static EMPTY: Vec<Vec<u32>> = Vec::new();
        self.path_cache
            .get(&(src, dst))
            .map(|v| v.as_slice())
            .unwrap_or(&EMPTY)
    }

    pub fn precompute_paths(&mut self, k: usize, max_hops: usize) {
        self.path_cache.clear();
        let n = self.vertices.len();
        for src in 0..n {
            for dst in 0..n {
                if src == dst {
                    continue;
                }
                let paths = self.yen_k_shortest(src as u32, dst as u32, k, max_hops);
                self.path_cache.insert((src as u32, dst as u32), paths);
            }
        }
    }

    fn yen_k_shortest(&self, src: u32, dst: u32, k: usize, max_hops: usize) -> Vec<Vec<u32>> {
        #[derive(Clone)]
        struct State {
            vertex: u32,
            path: Vec<u32>,
            cost: f64,
        }

        let mut result = Vec::new();
        let mut queue = VecDeque::new();
        queue.push_back(State {
            vertex: src,
            path: vec![],
            cost: 0.0,
        });

        while let Some(state) = queue.pop_front() {
            if result.len() >= k {
                break;
            }
            if state.vertex == dst {
                result.push(state.path);
                continue;
            }
            if state.path.len() >= max_hops {
                continue;
            }

            for (ei, edge) in self.edges.iter().enumerate() {
                if edge.src_vertex != state.vertex {
                    continue;
                }
                let cycle = state.path.iter().any(|&e| {
                    self.edges[e as usize].dst_vertex == edge.dst_vertex
                });
                if cycle {
                    continue;
                }
                let mut new_path = state.path.clone();
                new_path.push(ei as u32);
                let edge_cost = edge.attrs.latency_us + 1e6 / edge.attrs.bandwidth_gbps;
                queue.push_back(State {
                    vertex: edge.dst_vertex,
                    path: new_path,
                    cost: state.cost + edge_cost,
                });
            }
        }
        result
    }

    pub fn build_h100_node(node_id: u32, num_gpus: u32) -> Self {
        let mut g = Self::new();
        let gpu_verts: Vec<u32> = (0..num_gpus)
            .map(|i| {
                g.add_vertex(EndpointId {
                    node_id,
                    local_id: i,
                    endpoint_type: EndpointType::Gpu,
                })
            })
            .collect();

        let cpu = g.add_vertex(EndpointId {
            node_id,
            local_id: 0,
            endpoint_type: EndpointType::Cpu,
        });
        let cxl = g.add_vertex(EndpointId {
            node_id,
            local_id: 0,
            endpoint_type: EndpointType::CxlController,
        });
        let ib = g.add_vertex(EndpointId {
            node_id,
            local_id: 0,
            endpoint_type: EndpointType::IbPort,
        });

        let nvlink = FabricType::NvLink.default_attrs();
        let pcie = FabricType::Pcie.default_attrs();
        let cxl_attrs = FabricType::Cxl.default_attrs();
        let ib_attrs = FabricType::InfiniBand.default_attrs();

        for &i in &gpu_verts {
            for &j in &gpu_verts {
                if i != j {
                    g.add_edge(i, j, nvlink);
                }
            }
        }
        for &gv in &gpu_verts {
            g.add_edge(gv, cpu, pcie);
            g.add_edge(cpu, gv, pcie);
            g.add_edge(gv, cxl, cxl_attrs);
            g.add_edge(cxl, gv, cxl_attrs);
            g.add_edge(gv, ib, ib_attrs);
            g.add_edge(ib, gv, ib_attrs);
        }
        g.add_edge(cpu, ib, pcie);
        g.add_edge(ib, cpu, pcie);
        g
    }

    pub fn build_cluster(num_nodes: u32, gpus_per_node: u32) -> Self {
        let mut cluster = Self::new();
        let mut node_ib = Vec::new();

        for n in 0..num_nodes {
            let node = Self::build_h100_node(n, gpus_per_node);
            let base = cluster.vertices.len() as u32;
            for v in node.vertices {
                cluster.add_vertex(v);
            }
            for e in node.edges {
                cluster.add_edge(
                    e.src_vertex + base,
                    e.dst_vertex + base,
                    e.attrs,
                );
            }
            node_ib.push(base + gpus_per_node + 2);
        }

        let ib = FabricType::InfiniBand.default_attrs();
        for i in 0..num_nodes as usize {
            for j in (i + 1)..num_nodes as usize {
                cluster.add_edge(node_ib[i], node_ib[j], ib);
                cluster.add_edge(node_ib[j], node_ib[i], ib);
            }
        }
        cluster
    }
}

impl Default for TopologyGraph {
    fn default() -> Self {
        Self::new()
    }
}
